src/parser.rs

```diff
@@ -460,34 +461,74 @@ fn display_to_css_identifier<T: Display, W: fmt::Write>(x: &T, dest: &mut W) ->
     serialize_identifier(&string, dest)
 }
 
-fn compute_specificity<Impl: SelectorImpl>(mut selector: &ComplexSelector<Impl>,
-                                           pseudo_element: &Option<Impl::PseudoElement>) -> u32 {
-    struct Specificity {
-        id_selectors: u32,
-        class_like_selectors: u32,
-        element_selectors: u32,
+const MAX_10BIT: u32 = (1u32 << 10) - 1;
+
+struct Specificity {
+    id_selectors: u32,
+    class_like_selectors: u32,
+    element_selectors: u32,
+}
+
+impl Add for Specificity {
+    type Output = Specificity;
+
+    fn add(self, rhs: Specificity) -> Specificity {
+        Specificity {
+            id_selectors: self.id_selectors + rhs.id_selectors,
+            class_like_selectors:
+                self.class_like_selectors + rhs.class_like_selectors,
+            element_selectors:
+                self.element_selectors + rhs.element_selectors,
+        }
     }
-    let mut specificity = Specificity {
-        id_selectors: 0,
-        class_like_selectors: 0,
-        element_selectors: 0,
-    };
-    if pseudo_element.is_some() { specificity.element_selectors += 1 }
+}
 
-    compound_selector_specificity(&selector.compound_selector, &mut specificity);
-    loop {
-        match selector.next {
-            None => break,
-            Some((ref next_selector, _)) => {
-                selector = &**next_selector;
-                compound_selector_specificity(&selector.compound_selector, &mut specificity)
-            }
+impl Default for Specificity {
+    fn default() -> Specificity {
+        Specificity {
+            id_selectors: 0,
+            class_like_selectors: 0,
+            element_selectors: 0,
+        }
+    }
+}
+
+impl From<u32> for Specificity {
+    fn from(value: u32) -> Specificity {
+        assert!(value <= MAX_10BIT << 20 | MAX_10BIT << 10 | MAX_10BIT);
+        Specificity {
+            id_selectors: value >> 20,
+            class_like_selectors: (value >> 10) & MAX_10BIT,
+            element_selectors: value & MAX_10BIT,
         }
     }
+}
+
+impl From<Specificity> for u32 {
```

Comment:
Nit: I’m not sure about using the `From` trait here since this conversion is lossy, but maybe it doesn’t matter.